components/taxonomy/src/adapters/adapter.rs

```diff
@@ -0,0 +1,364 @@
+    ///
+    /// The signature's `accepts` represents the condition on values.
+    pub watch: Option<Signature>,
+}
+
+
+impl<'a> Feature<'a> {
+    /// Define an empty feature
+    pub fn empty(id: &Id<FeatureId>, service: &Id<ServiceId>) -> Self {
+        Feature {
+            id: id.clone(),
+            service: service.clone(),
+
+            implements: vec![],
+            tags: vec![],
+
+            send: None,
+            fetch: None,
+            delete: None,
+            watch: None,
+        }
+    }
+}
+
+
+/// The signature of a method in a `Feature`.
+#[derive(Clone)]
+pub struct Signature {
+    /// If `Nothing`, applying the method to the `Feature` never accepts an argument.
+    ///
+    /// If `Requires(format)`, applying the method to the `Feature` always requires an argument with
+    /// format `format`.
+    ///
+    /// If `Optional(format)`, applying the method to the `Feature` accepts an optional argument with
+    /// format `format`.
+    pub accepts: Expects<Arc<Format>>,
+
+    /// If `Nothing`, applying the method to the `Feature` never returns a result.
+    ///
+    /// If `Requires(format)`, applying the method to the `Feature` always returns a result with
+    /// format `format`.
+    ///
+    /// If `Optional(format)`, applying the method to the `Feature` may optionally return a result
+    /// with format `format`.
+    pub returns: Expects<Arc<Format>>,
+}
+impl Default for Signature {
+    fn default() -> Self {
+        Signature {
+            accepts: Expects::Nothing,
+            returns: Expects::Nothing,
+        }
+    }
+}
+impl Description for Signature {
+    fn description(&self) -> String {
+        format!("accepts: {}, returns: {}", self.accepts.description(), self.returns.description())
+    }
+}
+
+impl Description for Expects<Arc<Format>> {
+    fn description(&self) -> String {
+        match *self {
+            Expects::Nothing => "(nothing)".to_owned(),
+            Expects::Requires(ref format) => format.description(),
+            Expects::Optional(ref format) => format!("{} (optional)", format.description())
+        }
+    }
+}
+
+impl Description for Option<Signature> {
+    fn description(&self) -> String {
+        match *self {
+            None => "not supported".to_owned(),
+            Some(ref sig) => sig.description()
+        }
+    }
+}
+
+/// A witness that we are currently watching for a value.
+/// Watching stops when the guard is dropped.
+pub trait AdapterWatchGuard : Send + Sync {
+}
+
+pub enum WatchEvent {
+    /// Fired when we enter the range specified when we started watching, or if no range was
+    /// specified, fired whenever a new value is available.
+    Enter {
+        id: Id<FeatureId>,
+        value: Value
+    },
+
+    /// Fired when we exit the range specified when we started watching. If no range was
+    /// specified, never fired.
+    Exit {
+        id: Id<FeatureId>,
+        value: Value
+    }
+}
```

Comment:
I don't properly understand why we need Enter/Exit if we have no Range anymore now.